### PR TITLE
Fix ref analysis of self-assignment

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes https://github.com/dotnet/roslyn/issues/75592.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 1]
```diff
@@ -10170,5 +10170,78 @@ public void Utf8Addition()
...
+                {
+                    int field;
+                    ref int refField;
+
+                    public static void M([UnscopedRef] ref S s)
+                    {
+                        s.refField = ref s.field;
+                    }
+                }
+                """;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:10214]: Without the fix, this compiled without an error.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 2]
```diff
@@ -10170,5 +10170,78 @@ public void Utf8Addition()
...
+                S.M(ref s);
+
+                ref struct S
+                {
+                    int field;
+                    ref int refField;
+
+                    public static void M([UnscopedRef] ref S s)
+                    {
+                        s.refField = ref s.field;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:10211]: This is the code that should error as the _ref-safe-context_ is _return-only_.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs [hunk 3]
```diff
@@ -10170,5 +10170,78 @@ public void Utf8Addition()
...
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75592")]
+        public void SelfAssignment_CallerContext()
+        {
+            var source = """
+                using System.Diagnostics.CodeAnalysis;
+
+                S s = default;
+                S.M(ref s);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs:10202]: Consider that these two forms are equivalent 

```csharp
ref struct S {
  [UnscopedRef]
  void M1() { }

  static void M2([UnscopedRef] ref S p) { }
}

S local;
local.M1();
S.M2(ref local);
```

So if the `S.M(ref s)` is an error then virtually any instance method call on a `ref struct` with `[UnscopedRef]` is too..

**Review by jaredpar — CHANGES_REQUESTED** [review]
Think the error is in the wrong place here.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 4]
```diff
@@ -26477,6 +26477,12 @@ static ref S F2(S x2)
 
             var comp = CreateCompilation(source, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
             comp.VerifyEmitDiagnostics(
+                // (12,26): error CS8350: This combination of arguments to 'Program<S>.F1(ref S)' is disallowed because it may expose variables referenced by parameter 'x1' outside of their declaration scope
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:26480]: Can we add a variation of this test where the signature of `F1` is 

```csharp
static ref S F1(ref S x1)
```

**Review by jaredpar — APPROVED** [review]

**Pull request merged**